service/history/replication/task_processor.go

```diff
@@ -194,19 +194,18 @@ Loop:
 		respChan := make(chan *types.ReplicationMessages, 1)
 		// TODO: when we support prefetching, LastRetrievedMessageID can be different than LastProcessedMessageID
 
-		select {
-		case <-p.done:
+		if p.isShuttingDown() {
 			return
-		case p.requestChan <- &request{
+		}
+		p.requestChan <- &request{
```

Comment:
let's still leave the `select {}` block here with p.done check because we don't want this line to be blocking in case of shut down.

Reply:
ah, fair, I didn't consider this, let me fix